keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,354 @@
+	// 
+	// +optional
+	ControllerPublishSecretRef *LocalObjectReference
+
+	// NodeStageSecretRef is a reference to the secret object containing sensitive
+	// information to pass to the CSI driver to complete the CSI NodeStageVolume
+	// and NodeStageVolume and NodeUnstageVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	NodeStageSecretRef *LocalObjectReference
+
+	// NodePublishSecretRef is a reference to the secret object containing
+	// sensitive information to pass to the CSI driver to complete the CSI
+	// NodePublishVolume and NodeUnpublishVolume calls.
+	// This field is optional, and  may be empty if no secret is required. If the
+	// secret object contains more than one secret, all secret references are passed.
+	// 
+	// +optional
+	NodePublishSecretRef *LocalObjectReference
+}
+```
+
+### Secret references
+Secret references declared in an inline CSI volume can only be used with namespaces from pods where they are referenced .  For inline usage, secret references are stored in `LocalObjectReference` values:
+* `LocalObjectReference` do not include a namespace reference.  This is to prevent reference to arbitrary namespace values.
+* The namespace reference will be extracted from the the pod spec at different phases of storage lifecycle by the Kubelet or external CSI compoennts.
+* The Kubelet and external CSI components must ensure secret references can only be used with the namespace from inline pod spec.
+
+
+### Specifying allowed inline drivers with `PodSecurityPolicy`
+To control which driver is allowed to be used within a pod spec, this design will update the `PodSecurityPolicy` to introduce `AllowedCSIDrivers` as shown below:
+
+```go
+  type PodSecurityPolicySpec struct {
+	// <snip>
+
+	// AllowedCSIDrivers is a whitelist of allowed CSI drivers used inline in a pod spec.  Empty or nil indicates that all
+	// CSI drivers may be used.  This parameter is effective only when the usage of the CSI plugin
+	// is allowed in the "Volumes" field.
+	// +optional
+	AllowedCSIDrivers []AllowedCSIDriver
+  }
+
+  // AllowedCSIDriver represents a single CSI driver that is allowed to be used.
+  type AllowedCSIDriver struct {
+	// Name of the CSI driver
+	Name string
+  }  
+```
+
+Value `PodSecurityPolicy.AllowedCSIDrivers` must be explicitly set with the names of CSI drivers that are allowed to be embedded within a pod spec.  An empty value means no CSI drivers are allowed to be specified inline inside a pod spec.
+
+### Persistent vs ephemeral lifecycle setting
+A CSI driver must indicate that it supports either persistent or ephemeral inline mode.  This setting will be stored in a [`CSIDriver` configuration CRD](https://github.com/kubernetes/community/pull/2514) and must be retrieved from the API server at runtime.
```

Comment:
Given @kfox1111's clarification, I propose to change this into "A CSI driver must indicate that it supports ephemeral inline mode. This setting will be stored...".

What I'm missing in the document is an explicit statement that a driver can only be used in one mode or the other, and an explanation why that decision was made.

At first glance that decision seems to make it harder to deploy a driver that can operate in both modes. My colleagues are working on https://github.com/intel/pmem-csi, which manages local storage. Exposing that as ephemeral local storage is one way of using that storage, but we have ideas how we can make persistent mode also useful. It would be nice if we didn't have to run two driver daemons on each node just to support both. I can imagine doing it via two different CSI sockets provided by the same process, but if we could have a single socket and a parameter on `NodePublishVolume` for "this is an ephemeral volume" then that might be simpler.

Reply:
The issue is, Mount/Unmount of an ephemeral driver will do the creation of the backing store and the deletion of the backing store in the driver itself. The driver isn't handed any info to decide if it is an ephemeral driver or a normal one.

We might be able to relay this information in the future. I'd suggest though we make them totally separate for now and we can revisit having a driver that can do both modes in the future. We're really close to the deadline now.

Reply:
@pohly thanks for the feedback.  So I think a bit of clarification is in order (thanks for pointing that out). A driver should be able to support both at the same time. The lifecycle flag is consequential depending on how the volume is originated.  When volume is being used inline, in a pod, the lifecyle flag is used to determine how it should function.  When a volume is originated by a PV/PVC, it should automatically be assumed to be persistent.

The problem is, if a volume originates from PV/PVC and the driver does not in fact support persistent volume, then there will be issues.

Reply:
The protocol we came up with, for how an ephemeral driver will work does not have enough information to switch between being ephemeral and not being ephemeral.I say we only support one or the other explicitly for now, as this is alpha, and add that functionally later. perhaps as an enhancement to the pod info passing feature.

Reply:
FWIW, I'm fine with supporting just  one mode for now. I just wish the document would elaborate a bit more about the reasoning that led to the proposal. But that shouldn't stop it from getting merge as-is.

Reply:
@pohly I updated the doc to address the language and clarify your concerns.